### DISCUSSION TITLE
Ionization

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#512] [General]
Here is the deck i was trying to create out of the one mentioned in other post which fails to start. That is maximum what is possible to do reading other decks and 100 times the manuals. 

Where is the error and why it does not even start ?

```begin:control
  nx =  80
  ny =  160
  nz =  160

 nparticles = nx * ny * nz * 0.3

# Final time of simulation
  t_end = 200 * femto

# Size of domain
  x_min = -6 * micron
  x_max =  2 * micron
  y_min = -8 * micron
  y_max =  8 * micron
  z_min = -8 * micron
  z_max =  8 * micron

end:control

#-------------------------------------
begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = simple_outflow
  bc_y_max = simple_outflow
  bc_z_min = simple_outflow
  bc_z_max = simple_outflow
end:boundaries

#-----------------------------------------------------------
begin:constant
  laser_wavelength = 0.8 * micron
  omega = 2 * pi * c / laser_wavelength

  n_crit = critical(omega)
  # density (Fused Silica at full ionisation: 6.6e29)
  ne_max = 30 * n_crit

  laser_k = 2 * pi / laser_wavelength
  den_solid = 0.6875 * n_crit
  SiCharge = 1  # 12
  Ocharge  = 1  # 6

  ladder_step = -0.1 * micron

  r = sqrt(y^2 + z^2)

  laser_angle = pi / 4
  laser_fwhm = 5 * micron
  laser_dur_fwhm = 50.e-15

end:constant

begin:collisions
  use_collisions = T
  use_nanbu = T
  coulomb_log = auto
  collisional_ionisation = T
#  collide = electron electron on     #this key to switch off specific collisions because USE COLLISIONS=T means they are all ON by default (besides coll.ionization which needs additional switch to turn ON)
  coll_n_step = 10
  collide = all
end:collisions

#------------------------------------------------
begin:species
  name = Silicon
  charge = SiCharge  
  mass   = 1836.2*28.09
  ionise = T
  unique_electron_species = F

number_density = if( (x lt 0.3 * micron)    and (x gt 1. *ladder_step), den_solid, 0.)
number_density = if( (x lt 1. *ladder_step) and (x gt 2. *ladder_step), den_solid*0.9,  number_density(Silicon))
number_density = if( (x lt 2. *ladder_step) and (x gt 3. *ladder_step), den_solid*0.8,  number_density(Silicon))
number_density = if( (x lt 3. *ladder_step) and (x gt 4. *ladder_step), den_solid*0.7,  number_density(Silicon))
number_density = if( (x lt 4. *ladder_step) and (x gt 5. *ladder_step), den_solid*0.6,  number_density(Silicon))
number_density = if( (x lt 5. *ladder_step) and (x gt 6. *ladder_step), den_solid*0.5,  number_density(Silicon))
number_density = if( (x lt 6. *ladder_step) and (x gt 7. *ladder_step), den_solid*0.4,  number_density(Silicon))
number_density = if( (x lt 7. *ladder_step) and (x gt 8. *ladder_step), den_solid*0.3,  number_density(Silicon))
number_density = if( (x lt 8. *ladder_step) and (x gt 9. *ladder_step), den_solid*0.2,  number_density(Silicon))
number_density = if( (x lt 9. *ladder_step) and (x gt 10.*ladder_step), den_solid*0.1,  number_density(Silicon))
number_density = if( (x lt 10.*ladder_step) and (x gt 12.*ladder_step), den_solid*0.05, number_density(Silicon))


#  number_density = if( (x gt 0.) and (x lt target_x_outer), den_solid, 0.)
#  number_density = if( (x lt 0.) and (x gt target_x_wapor_inner), den_solid *(1+x/target_x_wapor_inner), number_density(Silicon))

   number_density_min = 0.05 * n_crit/SiCharge
#  density_max = ne_max
   npart_per_cell = 1
end:species
#---------------------------------------------------
begin:species
  name = Oxygen
  charge = Ocharge
  mass = 1836.2*16.09
  ionise = T
  unique_electron_species = F
  number_density = number_density(Silicon) * 2

  number_density_min = 0.05 * n_crit/Ocharge
# density_max = ne_max
  npart_per_cell = 1
end:species

#---------------------------------------------------
begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  number_density = number_density(Silicon)*SiCharge  + number_density(Oxygen)*Ocharge

  npart_per_cell = 1
  number_density_min = 0.01 * n_crit
  identify:electron
end:species

#--------------------------------------------------

begin:laser
  boundary = x_min

  intensity_w_cm2 = 2.0e18

  profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))
  t_profile = gauss(time, laser_dur_fwhm, laser_dur_fwhm / (2.0 * sqrt(loge(2.0))))

  phase = -y * laser_k * sin(laser_angle)          # Vary wave-front position in space
  lambda = laser_wavelength * cos(laser_angle)     # Space wave-fronts apart  # correctly along k

end:laser

#====================== O U T P U T =========================

begin:output_global
  force_final_to_be_restartable = T
end:output_global


begin:output
  name = normal
  file_prefix = x

# Simulated time between output dumps
  dt_snapshot = 20. * femto

# Properties on grid
  grid = always
  ey = always
#  charge = always
  average_particle_energy = always
  number_density = always + species
  temperature = always + species

  # Extended io
  distribution_functions = always
end:output ```

**Comment by Status-Mirror** [comment]
There are some errors here. Firstly you're missing the `begin:control` line, but I suspect this is a typo.

The main error is in the species blocks. When you run this code, you get the error message: 

```
 *** ERROR ***
 Ionising species must have an atomic number between
 1 and 100
```

As discussed in the ionisation section of the [species block documentation](https://epochpic.github.io/documentation/input_deck/input_deck_species.html), you must supply the code with an atomic number for it to determine which atom/ion species you are modelling. I have included a corrected input deck at the end of this message. If there is contradictory information in the documentation, let me know! This code runs, and my 4-core laptop estimates the job to take ~3 hours to run to completion.

```
begin:control 
  nx =  80
  ny =  160
  nz =  160

  nparticles = nx * ny * nz * 0.3

# Final time of simulation
  t_end = 200 * femto

# Size of domain
  x_min = -6 * micron
  x_max =  2 * micron
  y_min = -8 * micron
  y_max =  8 * micron
  z_min = -8 * micron
  z_max =  8 * micron

  stdout_frequency = 10

end:control

#-------------------------------------
begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = simple_outflow
  bc_y_max = simple_outflow
  bc_z_min = simple_outflow
  bc_z_max = simple_outflow
end:boundaries

#-----------------------------------------------------------
begin:constant
  laser_wavelength = 0.8 * micron
  omega = 2 * pi * c / laser_wavelength

  n_crit = critical(omega)
  # density (Fused Silica at full ionisation: 6.6e29)
  ne_max = 30 * n_crit

  laser_k = 2 * pi / laser_wavelength
  den_solid = 0.6875 * n_crit
  SiCharge = 1  # 12
  Ocharge  = 1  # 6

  ladder_step = -0.1 * micron

  r = sqrt(y^2 + z^2)

  laser_angle = pi / 4
  laser_fwhm = 5 * micron
  laser_dur_fwhm = 50.e-15

end:constant

begin:collisions
  use_collisions = T
  use_nanbu = T
  coulomb_log = auto
  collisional_ionisation = T
#  collide = electron electron on     #this key to switch off specific collisions because USE COLLISIONS=T means they are all ON by default (besides coll.ionization which needs additional switch to turn ON)
  coll_n_step = 10
  collide = all
end:collisions

#------------------------------------------------
begin:species
  name = Silicon
  charge = SiCharge  
  mass   = 1836.2*28.09
  ionise = T
  atomic_no = 14
  unique_electron_species = F

number_density = if( (x lt 0.3 * micron)    and (x gt 1. *ladder_step), den_solid, 0.)
number_density = if( (x lt 1. *ladder_step) and (x gt 2. *ladder_step), den_solid*0.9,  number_density(Silicon))
number_density = if( (x lt 2. *ladder_step) and (x gt 3. *ladder_step), den_solid*0.8,  number_density(Silicon))
number_density = if( (x lt 3. *ladder_step) and (x gt 4. *ladder_step), den_solid*0.7,  number_density(Silicon))
number_density = if( (x lt 4. *ladder_step) and (x gt 5. *ladder_step), den_solid*0.6,  number_density(Silicon))
number_density = if( (x lt 5. *ladder_step) and (x gt 6. *ladder_step), den_solid*0.5,  number_density(Silicon))
number_density = if( (x lt 6. *ladder_step) and (x gt 7. *ladder_step), den_solid*0.4,  number_density(Silicon))
number_density = if( (x lt 7. *ladder_step) and (x gt 8. *ladder_step), den_solid*0.3,  number_density(Silicon))
number_density = if( (x lt 8. *ladder_step) and (x gt 9. *ladder_step), den_solid*0.2,  number_density(Silicon))
number_density = if( (x lt 9. *ladder_step) and (x gt 10.*ladder_step), den_solid*0.1,  number_density(Silicon))
number_density = if( (x lt 10.*ladder_step) and (x gt 12.*ladder_step), den_solid*0.05, number_density(Silicon))


   number_density_min = 0.05 * n_crit/SiCharge
   npart_per_cell = 1
end:species
#---------------------------------------------------
begin:species
  name = Oxygen
  charge = Ocharge
  mass = 1836.2*16.09
  ionise = T
  atomic_no = 8
  unique_electron_species = F
  number_density = number_density(Silicon) * 2

  number_density_min = 0.05 * n_crit/Ocharge
# density_max = ne_max
  npart_per_cell = 1
end:species

#---------------------------------------------------
begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  number_density = number_density(Silicon)*SiCharge  + number_density(Oxygen)*Ocharge

  npart_per_cell = 1
  number_density_min = 0.01 * n_crit
  identify:electron
end:species

#--------------------------------------------------

begin:laser
  boundary = x_min

  intensity_w_cm2 = 2.0e18

  profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))
  t_profile = gauss(time, laser_dur_fwhm, laser_dur_fwhm / (2.0 * sqrt(loge(2.0))))

  phase = -y * laser_k * sin(laser_angle)          # Vary wave-front position in space
  lambda = laser_wavelength * cos(laser_angle)     # Space wave-fronts apart  # correctly along k

end:laser

#====================== O U T P U T =========================

begin:output_global
  force_final_to_be_restartable = T
end:output_global


begin:output
  name = normal
  file_prefix = x

# Simulated time between output dumps
  dt_snapshot = 20. * femto

# Properties on grid
  grid = always
  ey = always
#  charge = always
  average_particle_energy = always
  number_density = always + species
  temperature = always + species

  # Extended io
  distribution_functions = always
end:output
```

**Comment by DanRRRR** [comment]
Thanks Stuart,
1) I thought "name = Oxygen" completely defines atomic number like it did in very similar example but with no ionization in the other post #511... I can send you my Mendeleyev table from other codes if you like. It is doing the following (just needed to add more complete names of elements which i do not use) for element F

`	IF(name.EQ.'F'.OR.name.EQ.'f') atomic_no=9`

but better to convert all the names to lower case then it will look even easier
 
2) This example started to work, made couple sdf outputs but then by some reason crashed inside of my own code addition in the diagnostics.f90 which outputs all major PIC variables: 

**Program received signal SIGSEGV: Segmentation fault - invalid memory reference.**

2a) Do the code with ionization trying to performs some additional calls of "diagnostics" function compared to the case with no ionization? Because i output all needed variables i do not need it to do anything besides may be the option to restart the code for continuation. I removed  these lines too but this did not help

number_density = always + species
temperature = always + species

If the ionization calls diagnostics subroutine for its purposes what keywords it uses to distinguish itself from other calls? I will try to ignore these calls based on this keyword

2b) Before for many years I was  using completely different compiler and in my compiler when the error occurs i compile it with debugging switch and when the code crashes it starts the debugger and stops in it exactly on the offending line. How to debug this code in Linux?

**Reply by Status-Mirror** [reply]
The species name is arbitrary in EPOCH, it's simply a way to identify a given group. For example, we may wish to have an `oxygen_beam` species in the same simulation as an `oxygen_background`. 

As I said in #511, I don't believe there are any additional diagnostics calls with ionisation. Macro-particles will be added to the simulation, and macro-ions will change species - so make sure your diagnostics are able to handle a variable number of macro-particles in the output.

I assume there are proper debugging tools for parallel codes in Linux, but I personally just add in output statements until I find the offending line.

Cheers,
Stuart

**Comment by DanRRRR** [comment]
One more question about EPOCH ionization model: does the electron which ionizes ion lose the portion of its energy equal to ionization potential of ion? In other words, is total energy conserved ( kinetic energy of electron plus ionization energy of ion are the same before and after collisional ionization act) 

Looks like Segmentation crash i mentioned in previous message was caused by these lines, which may conflicted with my addition to the diagnostics subroutine which doing for me the output of essential PIC code particle variables (xm, pm, weight, and ID)
 
```  average_particle_energy = always
  number_density = always + species
  temperature = always + species

  # Extended io
  distribution_functions = always

**Reply by Status-Mirror** [reply]
Yes and no. Because we're dealing with macro-particles of different weights, we can't be guaranteed to conserve energy.

For example, a 1 keV electron may be expected to ionise 0.2 ions in a time-step. In the PIC code, the macro-electron may have a weight of 10, and the macro-ions may have a weight of 1000. Hence, our macro-electron is expected to create 2 real ions. Obviously if we ionise our 1000-weight macro-ion, we have created too many ions, but we can't just ignore the 2 ions that should have been created. Hence, we ionise the macro-ion with a probability of 2/1000, or 0.2%. Fantastic, problem solved - this is how EPOCH samples ionisations.

Now the energy conservation problem. If the ionisation occurs, we have a binding energy of 50 eV, and create a secondary electron with energy 100 eV, then there is an issue. The new macro-electron must have the same weight as the macro-ion to conserve charge, so this ionisation costs (1000 \* 100eV) + (1000 \* 50eV), or 150 keV. But our incident electron only has 1 keV of energy! If we want to sample the correct number of ions, then it's impossible to conserve energy.

Our solution is to decouple electron-energy loss from the ionisation sampling. If our 1 keV electron has a 20% chance of ionising an ion, then we will assume it has a 20% chance of losing energy in this step. If this 20% chance is met after sampling, then the energy it loses would be equivalent to ionising a macro-ion with the same weight as the macro-electron.

So **yes**, the algorithm conserves energy on average, but **no**, energy will not be conserved during a specific ionisation event. 

See  [my paper](https://pubs.aip.org/aip/pop/article-abstract/29/12/123907/2843764/Improvements-to-collisional-ionization-models-for?redirectedFrom=fulltext) for the full write-up on the collisional-ionisation algorithm.

**Comment by DanRRRR** [comment]
Stuart, 
One   strange thing: while i managed finally to start your example above, though did not yet see its results, the almost exactly  the same variant but just the 2D not 3D got me immediate error at loading:

```Name of species 1 is Silicon
 Name of species 2 is Silicon1
 Name of species 3 is Silicon2
 Name of species 4 is Silicon3
 Name of species 5 is Silicon4
 Name of species 6 is Silicon5
 Name of species 7 is Silicon6
 Name of species 8 is Silicon7
 Name of species 9 is Silicon8
 Name of species 10 is Silicon9
 Name of species 11 is Silicon10
 Name of species 12 is Silicon11
 Name of species 13 is Silicon12
 Name of species 14 is Silicon13
 Name of species 15 is Oxygen
 Name of species 16 is Oxygen1
 Name of species 17 is Oxygen2
 Name of species 18 is Oxygen3
 Name of species 19 is Oxygen4
 Name of species 20 is Oxygen5
 Name of species 21 is Oxygen6
 Name of species 22 is Oxygen7
 Name of species 23 is electron
 
 *** ERROR ***
 Missing release species for Silicon


Here is complete input deck 
--------------------------------------

begin:control 
  nx =  80
  ny =  160

# Final time of simulation
  t_end = 200 * femto

# Size of domain
  x_min = -6 * micron
  x_max =  2 * micron
  y_min = -8 * micron
  y_max =  8 * micron

  stdout_frequency = 10

end:control

#-------------------------------------
begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = simple_outflow
  bc_y_max = simple_outflow

end:boundaries

#-----------------------------------------------------------
begin:constant
  laser_wavelength = 0.8 * micron
  omega = 2 * pi * c / laser_wavelength

  n_crit = critical(omega)
  # density (Fused Silica at full ionisation: 6.6e29)
  ne_max = 30 * n_crit

  laser_k = 2 * pi / laser_wavelength
  den_solid = 0.6875 * n_crit
  SiCharge = 1  # 12
  Ocharge  = 1  # 6

  ladder_step = -0.1 * micron

  r = sqrt(y^2 + z^2)

  laser_angle = pi / 4
  laser_fwhm = 5 * micron
  laser_dur_fwhm = 50.e-15

end:constant

begin:collisions
  use_collisions = T
  use_nanbu = T
  coulomb_log = auto
  collisional_ionisation = T
#  collide = electron electron on     #this key to switch off specific collisions because USE COLLISIONS=T means they are all ON by default (besides coll.ionization which needs additional switch to turn ON)
  coll_n_step = 10
  collide = all
end:collisions

#------------------------------------------------
begin:species
  name = Silicon
  charge = SiCharge  
  mass   = 1836.2*28.09
  ionise = T
  atomic_no = 14
  unique_electron_species = F

number_density = if( (x lt 0.3 * micron)    and (x gt 1. *ladder_step), den_solid, 0.)
number_density = if( (x lt 1. *ladder_step) and (x gt 2. *ladder_step), den_solid*0.9,  number_density(Silicon))
number_density = if( (x lt 2. *ladder_step) and (x gt 3. *ladder_step), den_solid*0.8,  number_density(Silicon))
number_density = if( (x lt 3. *ladder_step) and (x gt 4. *ladder_step), den_solid*0.7,  number_density(Silicon))
number_density = if( (x lt 4. *ladder_step) and (x gt 5. *ladder_step), den_solid*0.6,  number_density(Silicon))
number_density = if( (x lt 5. *ladder_step) and (x gt 6. *ladder_step), den_solid*0.5,  number_density(Silicon))
number_density = if( (x lt 6. *ladder_step) and (x gt 7. *ladder_step), den_solid*0.4,  number_density(Silicon))
number_density = if( (x lt 7. *ladder_step) and (x gt 8. *ladder_step), den_solid*0.3,  number_density(Silicon))
number_density = if( (x lt 8. *ladder_step) and (x gt 9. *ladder_step), den_solid*0.2,  number_density(Silicon))
number_density = if( (x lt 9. *ladder_step) and (x gt 10.*ladder_step), den_solid*0.1,  number_density(Silicon))
number_density = if( (x lt 10.*ladder_step) and (x gt 12.*ladder_step), den_solid*0.05, number_density(Silicon))


   number_density_min = 0.05 * n_crit/SiCharge
   npart_per_cell = 1
end:species
#---------------------------------------------------
begin:species
  name = Oxygen
  charge = Ocharge
  mass = 1836.2*16.09
  ionise = T
  atomic_no = 8
  unique_electron_species = F
  number_density = number_density(Silicon) * 2

  number_density_min = 0.05 * n_crit/Ocharge

  npart_per_cell = 1
end:species

#---------------------------------------------------
begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  number_density = number_density(Silicon)*SiCharge  + number_density(Oxygen)*Ocharge

  npart_per_cell = 1
  number_density_min = 0.01 * n_crit
  identify:electron
end:species

#--------------------------------------------------

begin:laser
  boundary = x_min

  intensity_w_cm2 = 2.0e18

  profile = gauss(r, 0, laser_fwhm * (2.0 * sqrt(loge(2.0))))
  t_profile = gauss(time, 2*laser_dur_fwhm, laser_dur_fwhm * (2.0 * sqrt(loge(2.0))))

  phase = -y * laser_k * sin(laser_angle)          # Vary wave-front position in space
  lambda = laser_wavelength * cos(laser_angle)     # Space wave-fronts apart  # correctly along k

end:laser

#====================== O U T P U T =========================

begin:output_global
  force_final_to_be_restartable = T
end:output_global


begin:output
  name = normal
  file_prefix = x

# Simulated time between output dumps
  dt_snapshot = 20. * femto

# Properties on grid
  grid = always
  ey = always


#  charge = always
#  average_particle_energy = always
#  number_density = always + species
#  temperature = always + species

  # Extended io
#  distribution_functions = always
end:output

**Reply by Status-Mirror** [reply]
Hey @DanRRRR,

Looking at the error message:

```
 *** ERROR ***
 Missing release species for Silicon
```

I can see that you're missing a release species for Silicon. When you have ionisation, your ion moves to a species with a charge state of +1, and an electron is released. Currently, you haven't told EPOCH which species the release electron will go to. Look up the `ionisation_electron_species` key in the [species block](https://epochpic.github.io/documentation/input_deck/input_deck_species.html) documentation.

Alternatively, you could set `unique_electron_species = F`, and automatically generate a separate electron species to house released electrons from each Silicon charge-state.

Hope this helps,
Stuart

**Comment by DanRRRR** [comment]
Thanks Stuart. 
But how it happened that absolutely the same 3D case above did not complain and started running ? Results in 3D will be wrong?

**Reply by Status-Mirror** [reply]
I'd expect the same error message in 3D. I would need to see your 3D input deck to comment on the accuracy.

**Reply by DanRRRR** [reply]
I just noticed that 
unique_electron_species = F
was present in your 3D and mine 2D examples nevertheless the 2D run crashed.  

So, do i correctly understand that the 
"ionisation_electron_species" key  was missing in your example? And in simplest general case it has to be set "ionisation_electron_species=electron", correct ?

How about "unique_electron_species = F" -- is it still needed?

The goal is just to get classic process of ionization with degrees of ionization of each ion to follow the dynamics of plasma and electron population to follow the process of ionization. Initially when you do not feel how the code actually treats the ionization  and handles all the particles in the process few simplest working input.decks cases ideally with the comments are important to have.

**Comment by Status-Mirror** [comment]
Hey @DanRRRR,

I mananged to find some time to put together a page of [ionisation demos](https://epochpic.github.io/quickstart/basic_examples/ionisation.html). Let me know if you run into any problems with these simple test cases.

Cheers,
Stuart

**Comment by DanRRRR** [comment]
Thanks Stuart, i already saw that but had no time to investigate. The way EPOCH makes output of ionization data will cause problems with my codes. What it missing is the general way of output of ionization data without details what caused what.  In short as an example -- EPOCH starts with two species, electrons and say gold and at the end can create additionally 78 (!!!) new species instead of just still having these two species with different Z. Mostly no one cares how they were created to give to new ionization stage some special name (if you include recombination for example then there will be huge amount of back and forth ionizations and recombinations so there often is no sense to know how and what created this particular ion to justify special name for it - just its degree of ionization is enough to know for 99.999% cases. With ID you can follow each particular ion with no problem).

78 is just the start. If you include photorecombination then there will be 78 more. If include three-body recombination 78 additionally. Auto-ionization will add its 78. The code will stop running and will never end a single step

We discussed this with Chris couple years back and he suggested how  to implement output in this simple way --  when you know about species only their initial name and degree of ionization --  but noted that this   method was not debugged in his time when he was actively developing the code.

Additionally there exist some way to speedup the code with ionization sometimes by an order of magnitude. We will discuss that when i will finally make through existing method of output

**Reply by Status-Mirror** [reply]
I think there are some misconceptions in the ionisation routines from your message

- 78 additional species is just the start, but it's also the end. Once we have a species for each charge-state, we don't need to repeat this effort. Recombination could move particles about these pre-existing species. It doesn't matter about the process which produces ionisation, it's just important that each charged particle has its own species.
- The "per-particle-charge-mass" method you refer to will actually slow down the code. Each particle must store its own mass and charge variables in this system, so each time we do an MPI transfer of particles across ranks, that's an extra 2 variables per particle. The reason we have species with constant charge is so that we don't need to transfer this data.
- There's also a misconception that having multiple species will be slow. This is not always the case. The bottleneck of a PIC code is the number of particles, and we'd have the same number if we put them all in one species as we would have if we distributed them amongst the different species. It would, however, slow down ion-ion collisions - if there is sufficient interest in ion-ion collisions in systems with many ion-charge-states, then it might be worth ammending the collisions algorithm to combine all ions into a single group, and perform collisions between these. Generally I find people are more interested in electron motion than ion motion though.
